src/coreclr/jit/lclvars.cpp

```diff
@@ -1800,157 +1774,245 @@ bool Compiler::StructPromotionHelper::CanPromoteStructType(CORINFO_CLASS_HANDLE
-//   As a workaround for performance critical corner case: struct with 1 gcref, we try to construct
-//   the field information from indirect observations.
+//   true if the struct type can be promoted.
 //
-bool Compiler::StructPromotionHelper::CanConstructAndPromoteField(lvaStructPromotionInfo* structPromotionInfo)
+var_types Compiler::StructPromotionHelper::TryPromoteValueClassAsPrimitive(CORINFO_TYPE_LAYOUT_NODE* treeNodes,
+                                                                           size_t                    maxTreeNodes,
+                                                                           size_t                    index)
 {
-    const CORINFO_CLASS_HANDLE typeHnd    = structPromotionInfo->typeHnd;
-    const COMP_HANDLE          compHandle = compiler->info.compCompHnd;
-    const DWORD                typeFlags  = compHandle->getClassAttribs(typeHnd);
-    if (structPromotionInfo->fieldCnt != 1)
+    assert(index < maxTreeNodes);
+    CORINFO_TYPE_LAYOUT_NODE& node = treeNodes[index];
+    assert(node.type == CORINFO_TYPE_VALUECLASS);
+
+    if (node.simdTypeHnd != NO_CLASS_HANDLE)
     {
-        // Can't find out values for several fields.
-        return false;
+        const char* namespaceName = nullptr;
+        const char* className = compiler->info.compCompHnd->getClassNameFromMetadata(node.simdTypeHnd, &namespaceName);
+
+#ifdef FEATURE_SIMD
+        if (compiler->usesSIMDTypes() &&
+            (compiler->isRuntimeIntrinsicsNamespace(namespaceName) || compiler->isNumericsNamespace(namespaceName)))
+        {
+            unsigned    simdSize;
+            CorInfoType simdBaseJitType = compiler->getBaseJitTypeAndSizeOfSIMDType(node.simdTypeHnd, &simdSize);
+            // We will only promote fields of SIMD types that fit into a SIMD register.
+            if (simdBaseJitType != CORINFO_TYPE_UNDEF)
+            {
+                if (compiler->structSizeMightRepresentSIMDType(simdSize))
+                {
+                    return compiler->getSIMDTypeForSize(simdSize);
+                }
+            }
+        }
+#endif
+
+#ifdef TARGET_64BIT
+        // TODO-Quirk: Vector64 is a SIMD type with one 64-bit field, so when
+        // compiler->usesSIMDTypes() == false, it used to be promoted as a long
+        // field.
+        if (compiler->isRuntimeIntrinsicsNamespace(namespaceName) && (strcmp(className, "Vector64`1") == 0))
+        {
+            return TYP_LONG;
+        }
+#endif
     }
-    if ((typeFlags & CORINFO_FLG_CONTAINS_GC_PTR) == 0)
+
+    // Check for a single primitive wrapper.
+    if (node.numFields != 1)
     {
-        // Can't find out type of a non-gc field.
-        return false;
+        return TYP_UNDEF;
     }
 
-    const unsigned structSize = compHandle->getClassSize(typeHnd);
-    if (structSize != TARGET_POINTER_SIZE)
+    if (index + 1 >= maxTreeNodes)
     {
-        return false;
+        return TYP_UNDEF;
```

Comment:
Does this mean we must have a partial layout?

Reply:
No, this could be a SIMD intrinsic type that is not supported to be promoted on this platform.